### DISCUSSION TITLE
TPC-H configuration HammerDB

### DISCUSSION THREAD

**Discussion by ajithkumarg987** [TPC-Council/HammerDB#335] [Q&A] [answered]
I perceive TPC-H little  puzzle compared to TPC-H on CPUs/Cores and CPU utilization in HammerDB envirnoment.

Here the test TPC-H(PostgreSQL)
------------------

 schema_tpcc.tcl
-----------------
diset tpch pg_scale_fact 30
diset tpch pg_num_tpch_threads 10 /*schema buid with Vusers as n+1*/
diset tpch pg_degree_of_parallel 2 /*
print dict
buildschema

Test_run.tcl
-----------
vu set 4 /* why cannot test more vu's*/

Question1:
---------
if vu increased in Test_run.tcl got an error like "Virtual User 1: FATAL: terminating connection due to administrator command" and postgresql restart.
(ab)In schema build Vuser's used  more than 4 i.e pg_num_tpch_threads and test run why can't user more vu's.?

here is the limit for vu wrt scale-factor as suggested as HammerDB document"
The number of Virtual Users is based upon the following table where each Stream is processed by a Virtual User in HammerDB"?
https://www.hammerdb.com/docs/ch11s02.html

(b)Here what is best way more number virtual users wrt CPU cores? 
Is Virtual users limited in TPC-H and CPU Utilization or cores  related to this Parameters
max_worker_processes
max_parallel_workers_per_gather
max_parallel_workers
max_parallel_maintenance_workers

Question2:
---------

As HammerDB document as "For PostgreSQL the Degree of Parallelism sets the max_parallel_workers_per_gather parameter for the sessions executing the queries"
if the parameters max_parallel_workers_per_gather set in postgres.conf can be ignored to set in schema build i.e pg_degree_of_parallel. 
Mostly PostgreSQL read from the configuration.

**Comment by sm-shaw** [comment, marked as answer]
For Q1. pg_num_tpch_threads this is the number of virtual users to build the TPC-H schema.  It will automatically set the number of virtual users for the build based on this value, you do not need to use vu set to set it manually.  This can be any value you wish however you should not set it higher than the core count and also note that the virtual users will create and insert data into the same tables at the same time, therefore you may find setting this value too high can reduce the build time. Only testing will show what is right for your environment. 

For Q2.     This is what HammerDB runs (look at the driver script)
`set result [ pg_exec $lda "set max_parallel_workers_per_gather=$degree_of_parallel"]`
and here is the PostgreSQL documentation for what this setting does:
https://www.postgresql.org/docs/current/how-parallel-query-works.html
HammerDB is completely open source and transparent, you can always look at the code to see exactly what it is doing. This setting has the same impact as running the same at the command line.

**Reply by ajithkumarg987** [reply]
Thank you @sm-shaw  for answering to my questions.  still I've few more question on TPC-H . it helps me better understanding on tpc-h
Question3:
Perform the test.
#vuset vu 4 #Configure the number of concurrent tasks.
#vuset showoutput 1
#vuset logtotemp 1
#vuset unique 1
#print vuconf

Vuset --> the value should set lesser than CPU core count or the value set as "stream value based on scale factor"?
https://www.hammerdb.com/docs/ch11s02.html


Question4:
In TPC-C easily configured to run a particular Virtual User like 56 vu or 64 vu as show in below 

diset tpcc my_rampup 2
diset tpcc my_duration 5
vuset logtotemp 1
loadscript
puts " STARTED"
foreach z {56} /* only56 vu-configured*/
puts "$z VU TEST"
vuset vu $z
vucreate
vurun
runtimer 600
vudestroy

But In TPC-H how to configure a particular virtual user to run ?

Question6:Manually running Query's(Q1to Q22)in psql command line got an  error like invalid arguments for few quires,
SQL(2)/*TPCH QUERY GENERATION in //HammerDB-4.4/src/postgresql/pgolap.tcl
psql -U postgres  -d tpch -c "select s_acctbal, s_name, n_name, p_partkey, p_mfgr, s_address, s_phone, s_comment from part, supplier, partsupp, nation, region where p_partkey = ps_partkey and s_suppkey = ps_suppkey and p_size = :1 and p_type like '%:2' and s_nationkey = n_nationkey and n_regionkey = r_regionkey and r_name = ':3' and ps_supplycost = ( select min(ps_supplycost) from partsupp, supplier, nation, region where p_partkey = ps_partkey and s_suppkey = ps_suppkey and s_nationkey = n_nationkey and n_regionkey = r_regionkey and r_name = ':3') order by s_acctbal desc, n_name, s_name, p_partkey"


Question7:
TPC-H improvements over HammerDB-4.2vs4.3vs4.4?
Example same schema build with HammerDB-4.3vs HammerDBv4.4
/dev/ssd   /usr/postgres/data -->HammerDB-4.2-size 50GB.
/dev/ssd   /usr/postgres/data -->HammerDB-4.4-size 82GB.

**Reply by sm-shaw** [reply]
Hi, 
Q4. You should think of the Virtual Users differently between TPROC-C and TPROC-H. For TPROC-H (derived from TPC-H) there is the potential for 1 Virtual User to use all system resources with parallel query. Therefore, think of the approach as power test - one virtual user and throughput test, stream value based on scale factor, you set these with vuset. (There is no monitor virtual user). 
Q6. Yes, you cannot just extract the query from the source code, as the query predicates are generated at run time. To extract a query as it ran to re-run from the PostgreSQL command line run the HammerDB client with verbose output and out to the logfile and then run the actual query captured in the logfile. 
Q7. All the changes to HammerDB are recorded in the Changelog. These refer to the pull requests for you to see the code changes made.

**Reply by ajithkumarg987** [reply]
TPC-H still questionable compared to TPC-H on CPU utilization.
On go through the HammerDB documents not helpout on CPU uitlization. 

Example:
schema.tcl
diset tpch pg_scale_fact 30
diset tpch pg_num_tpch_threads 10
diset tpch degree_of_parallel/*max_parallel_workers_per_gather*/

test_run.tcl
vu set ->4

Question 1:
TPC-H:CPU utilization calculated on  scale_factor 30 on enabled parallel Querry(below configuration parameters)

max_worker_processes-> No of CPU's( total number of workers to the number of CPU cores)
max_parallel_workers=same for the number of parallel workers i.e  number of CPU's 
**max_parallel_workers_per_gather= workers an executor i.e the parallel execution of a planner node(Join/Hash/Merge)?
max_parallel_maintenance_work=No of  cores like physical cores(socket(s):2)  or Numa nodes(s):2.?**

(a)is PostgreSQL has an internal logic to calculate the maximum number of parallel workers it is ready to use for a table
max_parallel_workers_per_gather=log3(table size / min_parallel_table_scan_size(8MB)

CPU(s):              256
On-line CPU(s) list: 0-255
Thread(s) per core:  2
Core(s) per socket:  64
Socket(s):           2
NUMA node(s):        2

max_worker_processes=256
max_parallel_workers=256
**max_parallel_workers_per_gather=?/*log3(table size / min_parallel_table_scan_size(8MB)*/
max_parallel_maintenance_work?**?
** Total Number of CPU's/Process utilized on Scale factor 30?**

**Reply by sm-shaw** [reply]
You are out of scope for HammerDB support.  HammerDB TPROC-H is working as expected. Your questions are focused on how PostgreSQL manages parallel query and should be directed to a PostgreSQL discussion forum.

**Reply by ajithkumarg987** [reply]
How to  evaluate the TPC-H results on Multiple test runs on same configuration and schema
(a)is it may be better to compare the execution times of the queries individually?
Example:
Vuser 1:Executing Query 22 (12 of 22)
Vuser 1:query 22 completed in 0.999 seconds

(b)better to compare the execution time of completed Query sets/Geometric mean of query
user 1:Completed 1 query set(s)(Q1 to Q22) in 99 seconds
Vuser 1:Geometric mean of query times returning rows (22) is 0.99999

In (a) noticed the deviation is high(>20%)  between multiple test runs  for few query's.
In (b) noticed the deviation is low(<3%) btw multiple test run.

**Reply by sm-shaw** [reply]
Yes, you can compare the query times individually or use the geomean whichever works best for your environment. If query times are short then just comparing the times will be the best approach. The TPC prefers that you do not calculate the QphH metric yourself for derived workloads from TPC-H.
